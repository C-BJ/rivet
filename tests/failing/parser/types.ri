extern "C" fn malloc(sz: usize) *c_void;
extern "C" fn free(pointer: &c_void) c_void; // BAD

struct St {
    f1: &&i32;
    f2: *&i32;
    f3: &*i32;

    f4: ?*u8;

    f5: (i32, i32, i32, i32, i32, i32, i32, i32, i32);

    f6: ??i32;

    f7: c_void;

    f8: pkg;
    f9: =;
}

fn main() void { }
