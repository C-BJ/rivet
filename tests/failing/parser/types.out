tests/failing/parser/types.ri:1:32: error: cannot use `*c_void` as type
  1 | extern "C" fn malloc(sz: usize) *c_void;
    |                                ^
    = help: use `ptr` instead
tests/failing/parser/types.ri:2:29: error: cannot use references inside `extern` blocks
  2 | extern "C" fn free(pointer: &c_void) c_void; // BAD
    |                             ^
    = help: use pointers instead: `*c_void`
tests/failing/parser/types.ri:5:9: error: multi-level references are not allowed
  5 |     f1: &&i32;
    |         ^
tests/failing/parser/types.ri:6:9: error: cannot use pointers with references
  6 |     f2: *&i32;
    |         ^
tests/failing/parser/types.ri:7:9: error: cannot use references with pointers
  7 |     f3: &*i32;
    |         ^
tests/failing/parser/types.ri:9:9: error: pointers cannot be optional
  9 |     f4: ?*u8;
    |         ^
    = note: by default pointers can contain the value `none`
tests/failing/parser/types.ri:11:9: error: tuples can have a maximum of 8 types
  11 |     f5: (i32, i32, i32, i32, i32, i32, i32, i32, i32);
     |         ^
     = help: you can use a struct instead
tests/failing/parser/types.ri:13:9: error: optional multi-level types are not allowed
  13 |     f6: ??i32;
     |         ^
tests/failing/parser/types.ri:15:9: error: `c_void` can only be used inside `extern` declarations
  15 |     f7: c_void;
     |         ^
tests/failing/parser/types.ri:17:9: error: expected type, found keyword `pkg`
  17 |     f8: pkg;
     |         ^
tests/failing/parser/types.ri:18:9: error: expected type, found token `=`
  18 |     f9: =;
     |         ^
rivetc: error: could not compile package `types`, aborting due to 11 previous errors.